Chart builder in app: App initial create chart modal

cnathe opened the pull request:
#### Rationale
The app grids currently allow for viewing / rendering charts that have been saved for a given schema/query. However, currently these charts would have to be created in the LKS UI. This PR is step one in allowing creating, editing, and deleting charts for app grids (LKB only, experimental feature). This will allow for creating bar chart, scatter plot, box plot, pie chart, and line charts in the app. It does not include UI for changing any of the chart options at this time (i.e. axis labels, linear/log scale, colors, margins, etc.). 

#### Related Pull Requests
- https://github.com/LabKey/labkey-ui-components/pull/1456
- https://github.com/LabKey/labkey-ui-premium/pull/367
- https://github.com/LabKey/limsModules/pull/80
- https://github.com/LabKey/platform/pull/5364
- https://github.com/LabKey/testAutomation/pull/1868
- https://github.com/LabKey/tutorialModules/pull/181

#### Changes
- add experimental flag and update Chart menu for when to show Create Chart item
- ChartBuilderModal with modal for defining a new chart, selecting type, measures, and chart preview
- chart builder not to include grid filters in saved chart config
- save and delete chart from chart builder modal and reload model charts after save
- set showChartMenu false for several grids
- reload Chart menu and rendered chart after save (for create, update, or delete)
- Issue 48575: Split renderChartSVG into queryChartData and generateChartSVG so that app can just re-generateChartSVG on window size change
- use SVG icon for R reports in charts menu

cnathe committed:
add experimental flag and update Chart menu for when to show Create Chart item

cnathe committed:
remove hideEmptyChartMenu prop from GridPanel

cnathe committed:
fix GridPanel.spec.tsx

cnathe committed:
ChartBuilderMenuItem with initial chart builder modal display
- chart type, field selection inputs, and render chart preview

cnathe committed:
release notes update

cnathe committed:
3.28.0-fb-chartBuilderCreateModal.0

cnathe committed:
Chart builder modal to show loading spinner for preview while loading data
- and show preview message for some chart types when hiding points or binning

cnathe committed:
factor out getChartPreviewMsg as function

cnathe committed:
set showChartMenu false for APIKeysPanel grid

cnathe committed:
null check

cnathe committed:
jest fixes to use mountWithServerContext

cnathe committed:
3.28.0-fb-chartBuilderCreateModal.1

cnathe committed:
Allow save chart via Chart Builder modal
- use actions to loadCharts and set selected report after save

cnathe committed:
use new scatter plot and box plot icons in App charts menu

cnathe committed:
update release notes

cnathe committed:
3.28.0-fb-chartBuilderCreateModal.2

cnathe committed:
Move ChartBuilderMenuItem.tsx to /internal/components/chart

cnathe committed:
Chart loading mask min-height for chart grid display and the chart builder modal preview

cnathe committed:
Issue 48575: Split renderChartSVG into queryChartData and generateChartSVG so that app can just re-generateChartSVG on window size change

cnathe committed:
3.28.0-fb-chartBuilderCreateModal.3

cnathe committed:
don't save chart width/height and preview maxRows with chart config on app saveChart

cnathe committed:
Handle rendering of chart errors and display messages

cnathe committed:
use SVG icon for R reports in charts menu

cnathe committed:
3.28.0-fb-chartBuilderCreateModal.4

cnathe committed:
Merge branch 'develop' into fb_chartBuilderCreateModal

# Conflicts:
#	packages/components/package-lock.json
#	packages/components/package.json
#	packages/components/releaseNotes/components.md

cnathe committed:
3.29.0-fb-chartBuilderCreateModal.0

cnathe committed:
VisualizationConfigModel type and use for response of fetchGenericChart

cnathe committed:
Update perm check for when to show Create Chart menu item

cnathe committed:
refactor ChartBuilderModal.tsx out of ChartBuilderMenuItem.tsx

cnathe committed:
ChartBuilderModal to use savedChartModel for Edit Chart case

cnathe committed:
refactor ChartBuilderModal.tsx out of Chart (missed from previous commit)

cnathe committed:
update @labkey/api package version

cnathe committed:
Handle bar chart y-axis aggregate method on save, and display the selected method in chart builder modal

cnathe committed:
3.29.0-fb-chartBuilderCreateModal.2

cnathe committed:
Merge branch 'develop' into fb_chartBuilderCreateModal

# Conflicts:
#	packages/components/package-lock.json
#	packages/components/package.json
#	packages/components/releaseNotes/components.md

cnathe committed:
3.30.0-fb-chartBuilderCreateModal.0

cnathe committed:
saveChart error display in ChartBuilderModal

cnathe committed:
3.30.0-fb-chartBuilderCreateModal.1

cnathe committed:
ChartBuilderModal to allow delete chart if user canDelete

cnathe committed:
Show success message in app for create, update, and delete chart

cnathe committed:
ChartBuilderModal fixes for charts saved for a given view with filters on the view

cnathe committed:
jest test fixes to use mountWithAppServerContext

cnathe committed:
3.30.0-fb-chartBuilderCreateModal.2

cnathe committed:
fix for computeFilterKey value

cnathe committed:
Merge branch 'develop' into fb_chartBuilderCreateModal

# Conflicts:
#	packages/components/package-lock.json
#	packages/components/package.json
#	packages/components/releaseNotes/components.md

cnathe committed:
3.31.0-fb-chartBuilderCreateModal.0

cnathe committed:
trim name before saveChart

cnathe committed:
Merge branch 'develop' into fb_chartBuilderCreateModal

# Conflicts:
#	packages/components/package-lock.json
#	packages/components/package.json
#	packages/components/releaseNotes/components.md

cnathe committed:
3.32.1-fb-chartBuilderCreateModal.0

cnathe committed:
App chart builder not to include grid filters in saved chart config

cnathe committed:
3.32.1-fb-chartBuilderCreateModal.1

cnathe committed:
App chart builder should include model base filters in saved config (i.e. source samples grid case)

cnathe committed:
3.32.1-fb-chartBuilderCreateModal.2

cnathe committed:
apply chart containerFilter from containerFilter() helper

cnathe committed:
ManageDropdownButton prop for showIcon, default true

cnathe committed:
UsersGridPanel not to show charts menu

cnathe committed:
ChartPanel to hide chart edit icon based on exp feature enabled/disabled

cnathe committed:
3.32.1-fb-chartBuilderCreateModal.3

cnathe committed:
saved chart filter array change revert

cnathe committed:
3.32.1-fb-chartBuilderCreateModal.4

cnathe committed:
Chart builder modal preview chart should render with filters even if they aren’t saved

cnathe committed:
R report error message text update

cnathe committed:
ChartMenu to filter charts based on the selected view

cnathe committed:
Chart builder modal preview height updates

cnathe committed:
3.32.1-fb-chartBuilderCreateModal.5

cnathe committed:
ChartBuilderModal.tsx only supports a single axis measure at this time, so if an array of measures just use the first

cnathe committed:
jest test for ManageDropdownButton

cnathe committed:
jest test for ChartBuilderMenuItem.tsx

cnathe committed:
use updated chart icons

cnathe committed:
remove unused icons

cnathe committed:
3.32.1-fb-chartBuilderCreateModal.6

cnathe committed:
fix for chart builder type selected state icon

cnathe committed:
3.32.1-fb-chartBuilderCreateModal.7

cnathe committed:
Merge branch 'develop' into fb_chartBuilderCreateModal

# Conflicts:
#	packages/components/package-lock.json
#	packages/components/package.json
#	packages/components/releaseNotes/components.md

cnathe committed:
3.34.0-fb-chartBuilderCreateModal.0

cnathe committed:
jest tests for ChartBuilderModal.tsx

cnathe committed:
jest tests for ChartMenu.tsx

cnathe committed:
linting

labkey-alan commented on packages/components/src/internal/actions.ts:
```diff
@@ -866,3 +866,50 @@ export function renameGridView(
     });
 }
+
+interface SaveChartResponse {
```
I would put all of these where they are used (in ChartBuilderModal, above the component), unless there is a good reason to have them in a shared location like this.

cnathe replied:
I moved saveChart and deleteChart and the related interfaces to a new /components/chart/actions.ts file

labkey-alan commented on packages/components/src/internal/components/buttons/ManageDropdownButton.tsx:
```diff
@@ -17,15 +17,24 @@ import React, { FC, ReactNode, useMemo } from 'react';
         () => (
             <span>
-                <i className="fa fa-bars" /> Manage
+                {showIcon && (
+                    <>
+                        <i className="fa fa-bars" /> Manage
+                    </>
+                )}
+                {!showIcon && 'Manage'}
```
```suggestion
                {showIcon && <i className="fa fa-bars margin-right" />}
                Manage
```

cnathe replied:
done

labkey-alan commented on packages/components/src/internal/components/buttons/ManageDropdownButton.test.tsx:
```diff
@@ -0,0 +1,21 @@
+describe('ManageDropdownButton', () => {
+    test('default props', () => {
+        const { container } = render(<ManageDropdownButton />);
+        expect(container).toMatchSnapshot();
```
Let's not make these snapshot tests, this component is simple enough to test the same scenarios without using snapshots.

cnathe replied:
I switched them away from snapshot tests. I do think that these very simple components are a valid case for snapshot tests though since it is very clear when looking at the simple DOM when something changes to determine if that is an expected change or not.

labkey-alan commented on packages/components/src/internal/components/chart/Chart.tsx:
```diff
@@ -289,9 +307,10 @@ RReport.displayName = 'RReport';
     }
-
-    return <SVGChart api={api} chart={chart} container={container} filters={filters} />;
+    return null;
```
Is this case even possible?

cnathe replied:
Not currently. We filter the items that can show in the app charts menu to just R reports and generic charts at this time, but if that was ever changed then other report types would have tried to render as SVGChart. This change just makes it explicit.

labkey-alan commented on packages/components/src/internal/components/chart/api.ts:
```diff
@@ -3,19 +3,22 @@ import { Ajax, Filter, Query, Utils } from '@labkey/api';
 import { buildURL } from '../../url/AppURL';
 
-import { VisualizationConfigModel } from './models';
+import {GenericChartModel} from './models';
```
```suggestion
import { GenericChartModel } from './models';
```

cnathe replied:
done (I'll also do a linting pass before merging)

labkey-alan commented on packages/components/src/internal/components/chart/models.ts:
```diff
@@ -3,6 +3,7 @@ import { Filter, Query } from '@labkey/api';
 
 import { AppURL } from '../../url/AppURL';
 import { SchemaQuery } from '../../../public/SchemaQuery';
+import {VisualizationGetResponse} from "@labkey/api/dist/labkey/query/Visualization";
```
```suggestion
import { VisualizationGetResponse } from "@labkey/api/dist/labkey/query/Visualization";
```

Looks like we need to explicitly export `VisualizationGetResponse` from `@labkey/api`, or fix this import if we are already exporting it.

cnathe replied:
good catch. looks like I might need to bump @labkey/api to export VisualizationGetResponse

cnathe replied:
nvm, it is already there under the `Visualization` namespace

cnathe committed:
Merge branch 'develop' into fb_chartBuilderCreateModal

# Conflicts:
#	packages/components/package-lock.json
#	packages/components/package.json
#	packages/components/releaseNotes/components.md

cnathe committed:
3.34.1-fb-chartBuilderCreateModal.0

labkey-alan commented on packages/components/src/internal/components/chart/ChartBuilderModal.tsx:
```diff
@@ -0,0 +1,648 @@
+    );
+});
+
+export const getChartRenderMsg = (chartConfig: ChartConfig, rowCount: number, isPreview: boolean): string => {
+    let msg = '';
+    let sep = '';
+    if (isPreview && rowCount === MAX_ROWS_PREVIEW) {
+        msg = 'The preview is being limited to ' + MAX_ROWS_PREVIEW.toLocaleString() + ' rows.';
+        sep = ' ';
+    }
+
+    if (chartConfig.renderType === 'line_plot' && rowCount > chartConfig.geomOptions.binThreshold) {
+        msg +=
+            sep +
+            ('The number of individual points exceeds ' +
+                MAX_POINT_DISPLAY.toLocaleString() +
+                '. Data points will not be shown on this line plot.');
+    } else if (chartConfig.renderType === 'scatter_plot' && rowCount > MAX_POINT_DISPLAY) {
+        msg +=
+            sep +
+            ('The number of individual points exceeds ' +
+                MAX_POINT_DISPLAY.toLocaleString() +
+                '. The data is now grouped by density.');
+    }
+
+    return msg === '' ? undefined : msg;
+};
```
```suggestion
export const getChartRenderMsg = (chartConfig: ChartConfig, rowCount: number, isPreview: boolean): string => {
    const msg = [];
    if (isPreview && rowCount === MAX_ROWS_PREVIEW) {
        msg.push(`The preview is being limited to ${MAX_ROWS_PREVIEW.toLocaleString()} rows.`);
    }

    if (chartConfig.renderType === 'line_plot' && rowCount > chartConfig.geomOptions.binThreshold) {
        msg.push(`The number of individual points exceeds ${MAX_POINT_DISPLAY.toLocaleString()}.`);
        msg.push('Data points will not be shown on this line plot.');
    } else if (chartConfig.renderType === 'scatter_plot' && rowCount > MAX_POINT_DISPLAY) {
        msg.push(`The number of individual points exceeds ${MAX_POINT_DISPLAY.toLocaleString()}.`);
        msg.push('The data is now grouped by density.');
    }

    return msg.length === 0 ? undefined : msg.join(' ');
};
```

If we use `.join` on an array we don't have to track the separator, and if we append sentences to the array we can keep the line length shorter per message which is easier to read IMO.

cnathe replied:
yup, much cleaner. done

labkey-alan commented on packages/components/src/internal/components/chart/ChartBuilderModal.tsx:
```diff
@@ -0,0 +1,648 @@
+                return hasMatchingType || isMeasureDimensionMatch;
+            })
+            .sort(naturalSortByProperty('caption'))
+            .map(col => ({ label: col.caption, value: col.fieldKey, data: col })) ?? []
```
Is this `?? []` correct here? If any on of these methods returned undefined we'd encounter an error due to the way we're chaining, so I feel like `?? []` can never be true.

cnathe replied:
yup, not actually needed here. removed

labkey-alan commented on packages/components/src/internal/components/chart/ChartBuilderModal.tsx:
```diff
@@ -0,0 +1,648 @@
+    );
+});
+
+export const getChartRenderMsg = (chartConfig: ChartConfig, rowCount: number, isPreview: boolean): string => {
```
I would prefer if the stuff used by the component above was defined above the component. We do this in some other places and it triggers a linter warning, not sure why it's not doing that here.

cnathe replied:
done

labkey-alan commented on packages/components/src/internal/components/chart/ChartBuilderModal.tsx:
```diff
@@ -0,0 +1,648 @@
+                                                value={fieldValues[field.name]?.value}
+                                            />
+                                        </div>
+                                        {/* TODO change this to a select input with predefined options*/}
```
Is this for a later PR or this PR?

cnathe replied:
later PR, just using this display of the aggregate function for the time being until the next story

labkey-alan commented on packages/components/src/internal/components/chart/ChartBuilderModal.tsx:
```diff
@@ -0,0 +1,648 @@
+                                            type="checkbox"
+                                            checked={shared}
+                                            onChange={onToggleShared}
+                                        />{' '}
```
```suggestion
                                        />
```
This extra space here doesn't seem necessary. I tested locally with removing it and the spacing seems appropriate without it.

cnathe replied:
removed

labkey-alan commented on packages/components/src/internal/components/chart/ChartBuilderModal.tsx:
```diff
@@ -0,0 +1,648 @@
+        try {
+            const response = await saveChart(_reportConfig);
+            setSaving(false);
+            onHide(
+                (savedChartModel ? 'Successfully updated chart: ' : 'Successfully created chart: ') +
+                    _reportConfig.name +
+                    '.'
+            );
```
```suggestion
            onHide(`Sucessfully ${savedChartModel ? 'updated' : 'created'} chart: ${_reportConfig.name}.`);
```

cnathe replied:
done

labkey-alan commented on packages/components/src/internal/components/chart/ChartBuilderModal.tsx:
```diff
@@ -0,0 +1,648 @@
+        setName(event.target.value);
+    }, []);
+
+    const onToggleShared = useCallback(_ => {
+        setShared(prev => !prev);
+    }, []);
```
```suggestion
    const onToggleShared = useCallback(() => {
        setShared(prev => !prev);
    }, []);
```

cnathe replied:
done

labkey-alan commented on packages/components/src/internal/components/chart/ChartBuilderModal.tsx:
```diff
@@ -0,0 +1,648 @@
+                setShared(savedChartModel.shared);
+
+                const measures = savedChartModel.visualizationConfig?.chartConfig?.measures || {};
+                const fieldValues_ = {};
+                Object.keys(measures).map(key => {
+                    let measure = measures[key];
+                    if (measure) {
+                        // Currently only supporting a single measure per axis (i.e. not supporting y-axis left/right)
+                        if (Utils.isArray(measure)) measure = measure[0];
+
+                        fieldValues_[key] = { label: measure.label, value: measure.name, data: measure };
+                    }
+                });
```
```suggestion
                const fieldValues_ = Object.keys(measures).reduce((result, key) => {
                    let measure = measures[key];
                    if (measure) {
                        // Currently only supporting a single measure per axis (i.e. not supporting y-axis left/right)
                        if (Utils.isArray(measure)) measure = measure[0];

                        result[key] = { label: measure.label, value: measure.name, data: measure };
                    }
                    return result;
                }, {});
```

We should be using reduce here instead of map, especially since we're ignoring the result of map, and not returning anything within the map function.

cnathe replied:
yup, makes sense. done.

labkey-alan commented on packages/components/src/internal/components/chart/ChartBuilderModal.tsx:
```diff
@@ -0,0 +1,648 @@
+    const [showConfirmDelete, setShowConfirmDelete] = useState<boolean>(false);
+    const [error, setError] = useState<string>();
+    const [previewMsg, setPreviewMsg] = useState<string>();
+    const [reportConfig, setReportConfig] = useState<Record<string, any>>();
```
Is `Record<string, any>` appropriate here? Is the report config so dynamic we have to use `any`?

cnathe replied:
I changed it to define `SaveReportConfig`

labkey-alan commented on packages/components/src/internal/components/chart/ChartBuilderModal.tsx:
```diff
@@ -0,0 +1,648 @@
+
+export interface ChartFieldInfo {
+    aggregate?: AggregateFieldInfo;
+    // allowMultiple?: boolean;
```
We should delete this or at least leave a comment that indicates why we're keeping a field commented out.

cnathe replied:
comment added

labkey-alan commented on packages/components/src/internal/components/chart/ChartBuilderModal.tsx:
```diff
@@ -0,0 +1,648 @@
+    savedChartModel?: GenericChartModel;
+}
+
+export const ChartBuilderModal: FC<Props> = memo(({ actions, model, onHide, savedChartModel }) => {
```
My general feedback for this component is that the body of it is pretty complex, it's doing a lot (because it's designed to do a lot) and I think would benefit from being split into a few smaller components that can encapsulate the logic for each part. Especially considering that we're going to be adding to this thing over time, and it will only get more complicated. The "easy" divisions I see would be to split it into:

1. SideBar
2. ConfigForm
3. Preview
4. Footer

But there could definitely be other ways to split this up. I don't think we need a new file for each component.

cnathe replied:
done

labkey-alan commented on packages/components/src/theme/charts.scss:
```diff
@@ -180,3 +183,68 @@
 }
+
+.chart-builder-modal {
+    width: calc(100% - 100px) !important;
```
I don't think `!important` is necessary here if you define a separate style above that looks like:

```scss
.modal-dialog.chart-builder-modal {
    width: calc(100% - 100px);
}
```

cnathe replied:
done

cnathe replied:
done

labkey-alan commented on packages/components/src/theme/charts.scss:
```diff
@@ -180,3 +183,68 @@
     text-decoration: none;
 }
+
+.chart-builder-modal {
```
This block of styles being nested means we're created *very* specifically scoped styles, and I am unsure that we actually need to do that. In general we really only should be nesting styles if we want the styles scoped very specifically (and IME we don't want that for most cases).

labkey-alan replied:
[Here's a blog post ](https://bradfrost.com/blog/post/sass-selectors-to-nest-or-not-to-nest/)that goes over to nest or not nest, and included a (non-scientific) survey of what people were doing, and the general consensus was "We started nesting, but over time we found out it was not worth it, and have since stopped". I agree with the consensus here, in particular I find it hard to find the styles I care about when looking at DOM elements (which is something I do often) if the styles are nested.

It also tends to require a bunch of re-writes for the CSS if you change the DOM structure, because now your nested styles rely on the exact same nesting of DOM elements. If you don't nest the styles, and use a flatter hierarchy then styles won't break if you change the DOM. I think this is particularly relevant for this component, which is new, and will likely only grow in complexity and have one or more DOM changes in the near future as we add more functionality.

labkey-alan reviewed (changes requested):
Overall it looks good, and in my minimal manual testing I didn't notice any isssues outside of the issue with Hot Module Reload, which you are already aware of.

cnathe committed:
code review feedback
- split ChartBuilderModal component into subcomponents
- move chart actions to chart/actions.ts
- remove some scss nesting

cnathe committed:
3.34.1-fb-chartBuilderCreateModal.1

labkey-alan reviewed (approved):
Looks good. I did not do a super thorough manual testing pass, I am assuming that someone else is going to do that. If not I can prioritize it on Monday.

cnathe commented:
> Looks good. I did not do a super thorough manual testing pass, I am assuming that someone else is going to do that. If not I can prioritize it on Monday.

Great. Thanks. Chris did the manual testing for this story. I'll get it merged today.

cnathe committed:
npm run lint-branch-fix

cnathe committed:
Update release notes with version number and release date

cnathe committed:
3.35.0
